Implement MP3 Tagger

cbj4074 opened the pull request:
Now that FLAC tagging has been implemented, the same must be done for MP3.

cbj4074 committed:
refact: Eliminate duplication

cbj4074 committed:
refact: Delete unused MP3 conversion code

Converting MP3 files to other formats isn't particularly useful
because MP3 is lossy and there are few practical use-cases for it,
excepting only instances in which a lossless source format is not
available.

If sufficient demand arises in the future, such capabilities could
readily be added, as it's conceivable that a user may want to
convert an archive of historical MP3s to OGG files, for example.

cbj4074 committed:
tests: Move bulk of tests into base class in preparation for MP3 tests

cbj4074 committed:
wip: Abstract FLAC tests and implementation to make way for MP3

Essentially, the aim here is to abstract the tests and implementations
that are shared across file formats (e.g., FLAC and MP3) such that they
are generic wherever possible and therefore extendable when support for
other formats becomes necessary.

cbj4074 committed:
fix: Add missing methods to interface

cbj4074 committed:
fix: Remove unused argument and erroneous call to input() method

cbj4074 committed:
tests: Update tag key names to reflect names used in standard tools

The mid3v2 library uses these names, which are probably more sensible,
i.e., "song" is less ambiguous than "title" (it's not obvious whether
that label refers to the album title or the song title).

cbj4074 committed:
fix: Replace CLI tagging tool for MP3 (from lame to mid3v2)

Lame cannot tag-in-place; it re-encodes the file, which is not only
wasteful of resources, but a lossy process that further degrades the
quality of the resultant audio.

So, while Lame will still be used for transcoding (converting),
the ever-capable mid3v2 CLI tool will be used for tagging MP3 files.

cbj4074 committed:
fix: Use protected convenience method for removing all tags

Also, fix regression related to changes made in
58deabbe3088a086c4cd5bd26f8e06502034ea02.

cbj4074 committed:
feat: Add ability to set system locale per-process instance

While I'm violating the 2nd law of TDD here, I'm positive that this
will be necessary (due to the fact that it is in the legacy version),
and I want to commit it before I lose track of it.

cbj4074 committed:
fix: Add missing exception import

cbj4074 committed:
tests: Add convenience method for removing all tags during testing

cbj4074 committed:
tests: Make testItCanRemoveArtworkFromFile() abstract

cbj4074 committed:
feat: Implement methods to make estItCanRemoveArtworkFromFile() pass

cbj4074 committed:
ci: Add mid3v2 package to Travis CI environment

cbj4074 committed:
tests: Continue fleshing-out tagging tests

cbj4074 committed:
fix: Implement setEnv() method consistently and implement tag removal

cbj4074 committed:
tests: Add missing tests and simplify property name

cbj4074 committed:
wip: Change james-heinrich/getid3 dependency temporarily to fix issue

The fix for https://github.com/JamesHeinrich/getID3/issues/121 is in
the master branch only, for the moment.

cbj4074 committed:
feat: Improve failure message for exception

cbj4074 committed:
tests: Fix/implement a number of test methods and disable one (for now)

cbj4074 committed:
tests: Add tests for mid3v2 CLI

cbj4074 committed:
tests: Remove unused junk

cbj4074 committed:
refact: Remove unnecessary setEnv() method

cbj4074 committed:
refact: Remove unused method

Artwork is now embedded via the writeArtwork() method, which renders
this method is obsolete.

cbj4074 committed:
refact: Fix how supported output formats are set and delete unused code

poppabear8883 commented on composer.json:
```diff
@@ -20,7 +20,7 @@
 		"symfony/process": "^4.2",
 		"symfony/dependency-injection": "^4.2",
-		"james-heinrich/getid3": "^1.9",
+		"james-heinrich/getid3": "dev-master",
```
I am assuming because of the Hotfix/Patch you mentioned in chat ?

cbj4074 replied:
Yes, exactly.

poppabear8883 commented on src/CliCommand/Mid3v2Command.php:
```diff
@@ -0,0 +1,89 @@
+    public function __construct()
+    {
+        if (!empty(getenv('MID3V2_BINARY'))) {
+            $this->binary = getenv('MID3V2_BINARY');
```
This doesn't have to be done in this PR, However something to consider. Better practice would be to create configuration file(s) and have those only get values from the `env` ... This way if an `env` variable has to change at some point for whatever reason, you only update the configuration files and not have to modify several implementations to update the env variable in `getenv()` ... Hopefully this makes sense.

cbj4074 replied:
It does make sense, but my intention here was to allow for exactly what you describe.

It seems that any environment variable values should be set in the "parent" project that includes this library, and not in this library itself. In other words, the parent project could use something like https://github.com/vlucas/phpdotenv to set `MID3V2_BINARY=/home/foo/custom-mid3v2-binary` and the existing code would "just work", because `phpdotenv` would set the environment to reflect the values in the `.env` file.

I can't see any reason to use a config file in the library itself beyond what is already being done in `phpunit.xml` to set these values during testing.

Am I missing something here?

poppabear8883 replied:
So what happens when you decide that the variable `MID3V2_BINARY` now should be named something like `MID3V5_BINARY` ? ... you will have to `modify`  this file (and any other files) to reflect that change, where as you could simple update a config file and it will filter down to all inheriting files.

Also taking note to the filename itself (I am assuming that the name `Mid3v2` is formmated in the way of `Mid3v<version>` like it supports `v2.*`  ... note sure if this is something that could change in the future or not.

However, regarding the variable name above, maybe that's not why you change the variable name, but rather later determine that it makes better sense to name it something different ?

cbj4074 replied:
I agree that it would be less than ideal to have to update that `MID3V2_BINARY` name in more than one place, should it ever need to be changed.

That current name is based only on the actual library/binary name, which is in turn named after the `ID3v2` specification. The v2 refers to the specification at http://id3.org/id3v2.3.0 , and not the version of Mutagen (which happens to be the library that contains the binary we're using, `mid3v2`).

I think I see your point, now. You're simply suggesting that we make that a config file that remains local to this library, and not something that we intend to "publish" into the "parent" project, right? If that's the case, then yes, I concur, and we can do that in a subsequent PR.

poppabear8883 replied:
Mostly basing my opinion on the common practice that Laravel has implemented with environment variables.

https://github.com/alexeymezenin/laravel-best-practices#do-not-get-data-from-the-env-file-directly

Also is suggested by Laravel Shift, here is a related post https://laracasts.com/discuss/channels/general-discussion/laravel-shift-upgrade-to-52-use-config-instead-of-env

cbj4074 replied:
We'll have to give this further consideration and not implement anything on a whim.

I say that for two reasons, primarily:

1. I don't want this package to be coupled to Laravel, which means that any generic solution would likely need an additional "bridge" package to make it work seamlessly with Laravel (or any other framework that we intend to support).
2. The `.env` file and `config()` helper seem like they belong in the parent framework/app, and not in this (or any other) "library" that is included as a dependency thereto. That's not to say that we shouldn't find a way to avoid "hard-coding" the environment variable names (e.g., `MID3V2_BINARY`) within the implementing classes; I'm just not sure that an env/config package of some kind solves that problem elegantly, given our specific use-case here.

Let's open a separate Issue and/or PR for that effort, regardless.

poppabear8883 reviewed (approved):
Couple comments, otherwise LGTM

cbj4074 committed:
fix: Forego using picture switch in favor of legacy artwork embedding

Only the newest versions of mid3v2 support the --picture switch,
whereas in older versions, it is necessary to target the APIC frame.
In the interest of supporting a wider variety of environments, at least
until specific binary versions are included as proper dependencies,
the legacy method has been employed.

cbj4074 committed:
tests: Update test to reflect changes to how artwork is embedded in MP3

cbj4074 committed:
fix: Eliminate duplicate input file

The changes in 47af69d3e78ffda8ecce83cca868874f1ee42362 eliminate the
need to call input() because the file to be tagged is now specified
after the image file (whereas it was specified before, prior to that
change).

cbj4074 committed:
ci: Add commands to install and build the latest mutagen version

cbj4074 committed:
ci: Fix amateur-hour

cbj4074 committed:
ci: Fix incorrect CWD
